clippy_lints/src/option_if_let_else.rs

```diff
@@ -0,0 +1,176 @@
+    ///
+    /// ```rust
+    /// # let optional: Option<u32> = Some(0);
+    /// let _ = if let Some(foo) = optional {
+    ///     foo
+    /// } else {
+    ///     5
+    /// };
+    /// ```
+    ///
+    /// should be
+    ///
+    /// ```rust
+    /// # let optional: Option<u32> = Some(0);
+    /// let _ = optional.map_or(5, |foo| foo);
+    /// ```
+    pub OPTION_IF_LET_ELSE,
+    style,
+    "reimplementation of Option::map_or"
+}
+
+declare_lint_pass!(OptionIfLetElse => [OPTION_IF_LET_ELSE]);
+
+/// Returns true iff the given expression is the result of calling Result::ok
+fn is_result_ok(cx: &LateContext<'_, '_>, expr: &'_ Expr<'_>) -> bool {
+    if_chain! {
+        if let ExprKind::MethodCall(ref path, _, &[ref receiver]) = &expr.kind;
+        if path.ident.name.to_ident_string() == "ok";
+        if match_type(cx, &cx.tables.expr_ty(&receiver), &paths::RESULT);
+        then {
+            true
+        } else {
+            false
+        }
+    }
+}
+
+/// If this expression is the option if let/else construct we're detecting, then
+/// this function returns Some(Option<_> tested against, the method to call on
+/// the object (map_or or map_or_else), expression if Option is Some,
+/// expression if Option is None). Otherwise, it returns None.
+fn detect_option_if_let_else<'a>(
+    cx: &LateContext<'_, '_>,
+    expr: &'a Expr<'_>,
+) -> Option<(String, String, String, String)> {
+    if_chain! {
+        if let ExprKind::Match(let_body, arms, MatchSource::IfLetDesugar{contains_else_clause: true}) = &expr.kind;
+        if arms.len() == 2;
+        if match_type(cx, &cx.tables.expr_ty(let_body), &paths::OPTION);
+        if !is_result_ok(cx, let_body);
+        if let PatKind::TupleStruct(path, &[inner_pat], _) = &arms[0].pat.kind;
+        if let PatKind::Binding(_, _, id, _) = &inner_pat.kind;
+        then {
+            let (some_body, none_body) = if match_qpath(path, &paths::OPTION_SOME) {
+                (arms[0].body, arms[1].body)
+            } else {
+                (arms[1].body, arms[0].body)
+            };
+            let some_body = if let ExprKind::Block(Block { stmts: statements, expr: Some(expr), .. }, _)
+                = &some_body.kind {
+                if let &[] = &statements {
+                    expr
+                } else {
+                    // Don't trigger if there is a return, break, or continue inside
+                    if statements.iter().any(|statement| {
+                        match &statement {
+                            Stmt { kind: StmtKind::Semi(Expr { kind: ExprKind::Ret(..), ..}), .. } => true,
+                            Stmt { kind: StmtKind::Semi(Expr { kind: ExprKind::Break(..), ..}), .. } => true,
+                            Stmt { kind: StmtKind::Semi(Expr { kind: ExprKind::Continue(..), ..}), .. } => true,
+                            _ => false,
+                        }
+                    }) {
+                        return None;
+                    }
```

Comment:
That is not enough, since there could be a nested `return`/... somewhere. You have to implement a [`Visitor`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/intravisit/trait.Visitor.html) where this is checked in `visit_expr`.